README.md
# React Simple Typewriter

> A simple react component for adding a nice typewriter effect to your project.

[![NPM](https://img.shields.io/npm/v/react-simple-typewriter.svg)](https://www.npmjs.com/package/react-simple-typewriter) [![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)![npm bundle size](https://img.shields.io/bundlephobia/min/react-simple-typewriter)![GitHub](https://img.shields.io/github/license/awran5/react-simple-typewriter)

<p align="center">
  <img src="./screenshot.gif" alt="screenshot" />
</p>
 
## Install

#### npm

```sh
npm i react-simple-typewriter
```

#### Yarn

```sh
yarn add react-simple-typewriter
```

---

## Usage

There are two ways to `Typewriter`:

## 1. Component

```jsx
import React from 'react'
import { Typewriter } from 'react-simple-typewriter'

const MyComponent = () => {
  return (
    <div className='App'>
      <Typewriter {/* Props */} />
    </div>
  )
}
```

### Component Props

| Prop             |       Type        | Options  | Description                                                                                |      Default       |
| ---------------- | :---------------: | -------- | ------------------------------------------------------------------------------------------ | :----------------: |
| `words`          |       array       | Required | Array of strings holding the words                                                         | `['Hello', '...']` |
| `typeSpeed`      |      number       | Optional | Character typing speed in Milliseconds                                                     |        `80`        |
| `deleteSpeed`    |      number       | Optional | Character deleting speed in Milliseconds                                                   |        `50`        |
| `delaySpeed`     |      number       | Optional | Delay time between the words in Milliseconds                                               |       `1500`       |
| `loop`           | number \| boolean | Optional | Control how many times to run. `0 \| false` to run infinitely                              |        `1`         |
| `cursor`         |      boolean      | Optional | Show / Hide a cursor                                                                       |      `false`       |
| `cursorStyle`    |     ReactNode     | Optional | Change the cursor style available if `cursor` is `enabled`                                 |        `\|`        |
| `cursorBlinking` |      boolean      | Optional | Enable cursor blinking animation                                                           |        `\|`        |
| `onLoopDone`     |     function      | Optional | Callback function that is triggered when loops are completed. available if `loop` is `> 0` |        `-`         |
| `onType`         |     function      | Optional | Callback function that is triggered while typing with typed words `count` passed           |        `-`         |
| `onDelay`        |     function      | Optional | Callback function that is triggered on typing delay                                        |        `-`         |
| `onDelete`       |     function      | Optional | Callback function that is triggered while deleting                                         |        `-`         |

---

### Usage Example

```jsx
import React from 'react'
import { Typewriter } from 'react-simple-typewriter'

const MyComponent = () => {

  const handleType = (count: number) => {
    // access word count number
    console.log(count)}
  }

  const handleDone = () => {
    console.log(`Done after 5 loops!`)
  }

  return (
    <div className='App'>
      <h1 style={{ paddingTop: '5rem', margin: 'auto 0', fontWeight: 'normal' }}>
        Life is simple{' '}
        <span style={{ color: 'red', fontWeight: 'bold' }}>
          {/* Style will be inherited from the parent element */}
          <Typewriter
            words={['Eat', 'Sleep', 'Code', 'Repeat!']}
            loop={5}
            cursor
            cursorStyle='_'
            typeSpeed={70}
            deleteSpeed={50}
            delaySpeed={1000}
            onLoopDone={handleDone}
            onType={handleType}
          />
        </span>
      </h1>
    </div>
  )
}
```

## 2. Hook

**BREAKING CHANGES v5.0.0** Hook now returns `text` along with some useful `flags`:

| Prop       |  Type   | Description                         |
| ---------- | :-----: | ----------------------------------- |
| `isType`   | boolean | Check if currently typing           |
| `isDelete` | boolean | Check if currently deleting         |
| `isDelay`  | boolean | Check if currently on delay         |
| `isDone`   | boolean | Check if all running loops are done |

```jsx
import { useTypewriter } from 'react-simple-typewriter'

const MyComponent = () => {
  /**
   * @returns
   * text: [string] typed text
   * NEW helper: {} helper flags
   */
  const [text, helper] = useTypewriter({
    /* Config */
  })

  /* Hook helper */
  const { isType, isDelete, isDelay, isDone } = helper

  return (
    <div className='App'>
      <span>{text}</span>
    </div>
  )
}
```

### Hook Config

| Prop          |       Type        | Options  | Description                                                                                |      Default       |
| ------------- | :---------------: | -------- | ------------------------------------------------------------------------------------------ | :----------------: |
| `words`       |       array       | Required | Array of strings holding the words                                                         | `['Hello', '...']` |
| `typeSpeed`   |      number       | Optional | Character typing speed in Milliseconds                                                     |        `80`        |
| `deleteSpeed` |      number       | Optional | Character deleting speed in Milliseconds                                                   |        `50`        |
| `delaySpeed`  |      number       | Optional | Delay time between the words in Milliseconds                                               |       `1500`       |
| `loop`        | number \| boolean | Optional | Control how many times to run. `0 \| false` to run infinitely                              |        `1`         |
| `onLoopDone`  |     function      | Optional | Callback function that is triggered when loops are completed. available if `loop` is `> 0` |        `-`         |
| `onType`      |     function      | Optional | Callback function that is triggered while typing                                           |        `-`         |
| `onDelete`    |     function      | Optional | Callback function that is triggered while deleting                                         |        `-`         |
| `onDelay`     |     function      | Optional | Callback function that is triggered on typing delay                                        |        `-`         |

### Hook Usage Example

```jsx
import React from 'react'
import { useTypewriter } from 'react-simple-typewriter'

const MyComponent = () => {
  const [text] = useTypewriter({
    words: ['Hello', 'From', 'Typewriter', 'Hook!'],
    loop: 0
  })

  return (
    <div className='App'>
      <span>{text}</span>
    </div>
  )
}
```

### Hook with Cursor

If you like to have the **Cursor** effect, you can `import` it as a separate `Component`

```jsx
import React from 'react'
import { useTypewriter, Cursor } from 'react-simple-typewriter'

const MyComponent = () => {
  const [text] = useTypewriter({
    words: ['Hello', 'From', 'Typewriter', 'Hook!'],
    loop: 3,
    onLoopDone: () => console.log(`loop completed after 3 runs.`)
  })

  return (
    <div className='App'>
      <span>{text}</span>
      <Cursor cursorColor='red' />
    </div>
  )
}
```

### Cursor Component Props

| Prop             |   Type    | Options  | Description                       |  Default  |
| ---------------- | :-------: | -------- | --------------------------------- | :-------: |
| `cursorStyle`    | ReactNode | Optional | Change cursor style               |   `\|`    |
| `cursorColor`    |  String   | Optional | Change cursor color               | `inherit` |
| `cursorBlinking` |  Boolean  | Optional | disable cursor blinking animation |  `true`   |

---

### [Demo](https://react-simple-typewriter.vercel.app/)

<br />

[![Edit react-simple-typewriter](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/react-simple-typewriter-uj8ix)

### License

MIT © [awran5](https://github.com/awran5/)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

## [v5.0.1](https://github.com/awran5/react-simple-typewriter/compare/v5.0.1...v5.0.1)

## [v5.0.1](https://github.com/awran5/react-simple-typewriter/compare/v5.0.0...v5.0.1) - 2022-11-19

### Commits

- docs: update [`8c889b7`](https://github.com/awran5/react-simple-typewriter/commit/8c889b7646238a889d59123b87926282d5b4a2a9)
- fix: reset `isDelay` flag when done [`1dab1d8`](https://github.com/awran5/react-simple-typewriter/commit/1dab1d813827eed842afe08db23368b79f66b0ee)

## [v5.0.0](https://github.com/awran5/react-simple-typewriter/compare/v4.0.5...v5.0.0) - 2022-11-19

### Commits

- build: save upgrade [`c67ab79`](https://github.com/awran5/react-simple-typewriter/commit/c67ab7960b1d6f925202d38920bddd7a93443e4f)
- docs: update docs [`04c42ca`](https://github.com/awran5/react-simple-typewriter/commit/04c42ca4b3e776856db42d4e3541c59afbbe0c6c)
- feat: refactor `state` and hook `output` [`64c63e7`](https://github.com/awran5/react-simple-typewriter/commit/64c63e7beaa08ed88248cadf7d7fd2a4802aed38)

## [v4.0.5](https://github.com/awran5/react-simple-typewriter/compare/v4.0.4...v4.0.5) - 2022-10-05

### Commits

- build: upgrade dependencies [`0470042`](https://github.com/awran5/react-simple-typewriter/commit/0470042415345ba0a1d7c5393a08cc7c81d58c1a)

## [v4.0.4](https://github.com/awran5/react-simple-typewriter/compare/v4.0.3...v4.0.4) - 2022-10-04

### Commits

- 4.0. [`870c281`](https://github.com/awran5/react-simple-typewriter/commit/870c281d1ac928a0e4ce172e1553d99cda988017)
- docs: add missing peer dependencies [`53ed9a4`](https://github.com/awran5/react-simple-typewriter/commit/53ed9a4d713d3b73c3fc7e3447adbc26fa007788)
- update [`b059c7f`](https://github.com/awran5/react-simple-typewriter/commit/b059c7f0d51eac8cb653a27a9c9fb0302ffdd027)

## [v4.0.3](https://github.com/awran5/react-simple-typewriter/compare/v4.0.2...v4.0.3) - 2022-10-04

### Commits

- build: remove unused dependencies [`1abdcd0`](https://github.com/awran5/react-simple-typewriter/commit/1abdcd03a39915feaf0add52f754601091118892)
- feat: remove chromatic [`5089fe0`](https://github.com/awran5/react-simple-typewriter/commit/5089fe02429b1d04717fefb12746745fd36bb3f6)
- chore: rename script [`edda0c0`](https://github.com/awran5/react-simple-typewriter/commit/edda0c0db77dd85b159f62396056410e04f2273a)

## [v4.0.2](https://github.com/awran5/react-simple-typewriter/compare/v4.0.1...v4.0.2) - 2022-09-08

### Commits

- chore: add chromatic [`f6c1967`](https://github.com/awran5/react-simple-typewriter/commit/f6c196773782febeed8818794b1983388a191064)
- docs: update readme [`cb503a1`](https://github.com/awran5/react-simple-typewriter/commit/cb503a16e522976e02efb481470a1fe0e73e5cd6)
- chore: update [`e2843fa`](https://github.com/awran5/react-simple-typewriter/commit/e2843fac14fb5087a3c2444ee82859b97eccf8db)

## [v4.0.1](https://github.com/awran5/react-simple-typewriter/compare/v4.0.0...v4.0.1) - 2022-09-08

### Commits

- docs: update [`f91bd35`](https://github.com/awran5/react-simple-typewriter/commit/f91bd35a76ef3ab251444d06073966ceb3a5730d)
- docs: update [`daaa240`](https://github.com/awran5/react-simple-typewriter/commit/daaa240dbf852549b42cd19ffa81aab5cccb080a)

## [v4.0.0](https://github.com/awran5/react-simple-typewriter/compare/v3.0.3...v4.0.0) - 2022-09-08

### Commits

- docs: add storybook new stories [`5dcf1c3`](https://github.com/awran5/react-simple-typewriter/commit/5dcf1c3d3adaeea03c7907cb775c569dbfcd8da8)
- refactor: add gh-release [`bbd993c`](https://github.com/awran5/react-simple-typewriter/commit/bbd993cbc6de7369e2a80eb6cc436c958348053f)
- docs: remove theme-ui mdx-embed [`105493d`](https://github.com/awran5/react-simple-typewriter/commit/105493dc93fe9409d4ce6aae3ccbf5c122ed4477)

## [v3.0.3](https://github.com/awran5/react-simple-typewriter/compare/v3.0.2...v3.0.3) - 2022-09-08

### Commits

- refactor: upgrade dependencies [`977fcd1`](https://github.com/awran5/react-simple-typewriter/commit/977fcd1f50bee27b8cd6074469f9454cdcc18177)
- refactor: added: storybook [`d667cff`](https://github.com/awran5/react-simple-typewriter/commit/d667cff6eca68ddee47c657e65450bfcb106b8c4)
- refactor: remove github action [`45ef186`](https://github.com/awran5/react-simple-typewriter/commit/45ef1865ebc6c3046c7c99b8fc294c81b640795a)

## [v3.0.2](https://github.com/awran5/react-simple-typewriter/compare/v3.0.1...v3.0.2) - 2022-09-08

### Commits

- refactor: remove auto versioning dependencies [`3711529`](https://github.com/awran5/react-simple-typewriter/commit/3711529849eae1feae0df3760d4e4e6da9ccbe86)
- refactor: remove eslint [`31f5ec7`](https://github.com/awran5/react-simple-typewriter/commit/31f5ec72c1ac99f94e0bc72f68c0038f2cc04b2f)
- refactor: remove unused package.json scripts [`f47a831`](https://github.com/awran5/react-simple-typewriter/commit/f47a831fade76d2398bd5988c4543335ddb918da)

## [v3.0.1](https://github.com/awran5/react-simple-typewriter/compare/v3.0.0...v3.0.1) - 2021-08-27

### Merged

- Expose count ref [`#3`](https://github.com/awran5/react-simple-typewriter/pull/3)

### Commits

- perf: updates and performance improved [`65b57c1`](https://github.com/awran5/react-simple-typewriter/commit/65b57c18a5eb76315a378068a00b64f795efd006)
- expose count ref to parent [`4cdbc6b`](https://github.com/awran5/react-simple-typewriter/commit/4cdbc6bc272558db356119c339d7c970a6188826)
- Update useTypewriter.tsx [`11996cc`](https://github.com/awran5/react-simple-typewriter/commit/11996cc4aa430ecc6fbe53f7020bbef7beb05418)

## [v3.0.0](https://github.com/awran5/react-simple-typewriter/compare/v2.0.4...v3.0.0) - 2021-07-06

### Commits


src/index.js
import React, { useState, useEffect, useCallback, memo, useRef } from 'react'
import PropTypes from 'prop-types'

function Typewriter({
  loop,
  typeSpeed,
  deleteSpeed,
  delaySpeed,
  words,
  cursor,
  cursorStyle
}) {
  // State
  const [speed, setSpeed] = useState(100)
  const [text, setText] = useState('')
  // Refs
  const isDeleting = useRef(false)
  const counter = useRef(0)

  const handleTyping = useCallback(() => {
    const index = loop ? counter.current % words.length : counter.current
    const word = words[index]
    setSpeed(typeSpeed)

    if (isDeleting.current) {
      // Set stoping speed
      setSpeed(deleteSpeed)
      // Move backwards
      setText((prev) => word.substring(0, prev.length - 1))
      // Move to the next word when text is empty
      if (text === '') {
        isDeleting.current = false
        counter.current = counter.current + 1
      }
    } else {
      // Move Forward
      setText((prev) => word.substring(0, prev.length + 1))
      // Word is completed
      if (text === word) {
        // if not loop just return
        if (!loop) return
        // else set the speed delay and start over
        setSpeed(delaySpeed)
        isDeleting.current = true
      }
    }
  }, [delaySpeed, deleteSpeed, loop, text, typeSpeed, words])

  useEffect(() => {
    const timer = setTimeout(() => handleTyping(), speed)
    return () => clearTimeout(timer)
  }, [handleTyping, speed])

  return (
    <React.Fragment>
      <span>{text}</span>
      {cursor && <span className='blinking-cursor'>{cursorStyle}</span>}
    </React.Fragment>
  )
}

Typewriter.propTypes = {
  typeSpeed: PropTypes.number,
  deleteSpeed: PropTypes.number,
  delaySpeed: PropTypes.number,
  words: PropTypes.array.isRequired,
  cursorStyle: PropTypes.string,
  cursor: PropTypes.bool,
  loop: PropTypes.bool
}

Typewriter.defaultProps = {
  typeSpeed: 100,
  deleteSpeed: 50,
  delaySpeed: 1500,
  words: ['Eat', 'Sleep', 'Code', 'Repeat!'],
  cursorStyle: '|',
  cursor: false,
  loop: false
}

export default memo(Typewriter)
